packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go

```diff
@@ -726,13 +950,93 @@ func (dhtPeer *DHTPeer) lookupAddressDHT(address string) (peer.ID, error) {
-			if err == nil {
-				connected = true
-				break
+			if err != nil {
+				lwarn().Str("op", "lookup").Str("addr", address).
+					Msgf("couldn't open stream to address provider %s: %s, looking up other providers...", provider, err.Error())
+				dhtPeer.routedHost.Peerstore().ClearAddrs(provider.ID)
+				continue
+			}
+
+			linfo().Str("op", "lookup").Str("addr", address).
+				Msgf("getting agent record from provider %s...", provider)
+
+			// prepare LookupRequest
+			lookupRequest := &dhtnode.LookupRequest{AgentAddress: address}
+			msg := &dhtnode.AcnMessage{Version: dhtnode.CurrentVersion, Payload: &dhtnode.AcnMessage_LookupRequest{LookupRequest: lookupRequest}}
+			buf, err := proto.Marshal(msg)
+			ignore(err)
+
+			err = utils.WriteBytes(stream, []byte(buf))
+			if err != nil {
+				lwarn().Str("op", "lookup").Str("addr", address).
+					Msgf("couldn't send agent lookup request to provider %s (%s), looking up other providers...", provider, err.Error())
+				err = stream.Reset()
+				ignore(err)
+				continue
+			}
+
+			buf, err = utils.ReadBytes(stream)
+			if err != nil {
+				lwarn().Str("op", "lookup").Str("addr", address).
+					Msgf("couldn't receive agent lookup response from provider %s (%s), looking up other providers...", provider, err.Error())
+				err = stream.Reset()
+				ignore(err)
+				continue
+			}
+
+			stream.Close()
+
+			response := &dhtnode.AcnMessage{}
+			err = proto.Unmarshal(buf, response)
+			if err != nil {
+				lwarn().Str("op", "lookup").Str("addr", address).
+					Msgf("couldn't deserialize agent lookup response from provider %s (%s), looking up other providers...", provider, err.Error())
+				continue
+			}
+
+			// Response is either a LookupResponse or Status
+			var lookupResponse *dhtnode.LookupResponse = nil
+			var status *dhtnode.Status = nil
+			switch pl := response.Payload.(type) {
+			case *dhtnode.AcnMessage_LookupResponse:
+				lookupResponse = pl.LookupResponse
+			case *dhtnode.AcnMessage_Status:
+				status = pl.Status
+			default:
+				err = errors.New("Unexpected Acn Message")
+				lwarn().Str("op", "lookup").Str("addr", address).
+					Msgf("couldn't deserialize agent lookup response from provider %s (%s), looking up other providers...", provider, err.Error())
+				continue
 			}
 
-			dhtPeer.routedHost.Peerstore().ClearAddrs(provider.ID)
-			lwarn().Str("op", "lookup").Str("addr", address).Msgf("couldn't open stream to address provider %s: %s, looking up for other providers...", provider, err.Error())
+			if status != nil {
+				err = errors.New(status.Code.String() + " : " + strings.Join(status.Msgs, ":"))
+				lwarn().Str("op", "lookup").Str("addr", address).
+					Msgf("Failed agent lookup response from provider %s (%s), looking up other providers...", provider, err.Error())
+				continue
+			}
+
+			// lookupResponse must be set
+			record := lookupResponse.AgentRecord
+			valid, err := dhtnode.IsValidProofOfRepresentation(record, address, record.PeerPublicKey)
```

Comment:
[STP3.a] when a peer receives outgoing envelope it proceed to continuously lookup the dht for the destination until it finds a peer with a valid PoR, then forward the envelope and wait for delivery confirmation